client/webserver/webserver.go

```diff
@@ -398,11 +410,105 @@ func (s *WebServer) isAuthed(r *http.Request) bool {
 		log.Errorf("authToken retrieval error: %v", err)
 	}
 
+	return authToken
+}
+
+// getPWKey checks the request for a password key cookie. Returns an error
+// if it does not exist or it is not a valid.
+func getPWKey(r *http.Request) ([]byte, error) {
+	cookie, err := r.Cookie(pwKeyCK)
+	switch {
+	case err == nil:
+		sessionKey, err := hex.DecodeString(cookie.Value)
+		if err != nil {
+			return nil, err
+		}
+		return sessionKey, nil
+	default:
+		return nil, err
+	}
+}
+
+// isAuthed checks if the incoming request is from an authorized user/device.
+// Requires the auth token cookie to be set in the request and for the token
+// to match `WebServer.validAuthToken`.
+func (s *WebServer) isAuthed(r *http.Request) bool {
+	authToken := getAuthToken(r)
+	if authToken == "" {
+		return false
+	}
 	s.mtx.RLock()
 	defer s.mtx.RUnlock()
 	return s.authTokens[authToken]
 }
 
+// getCachedPassword takes authToken passed to and the keyBlob returned from
+// cacheAppPassword to retrieve and decrypt the app password.
+func (s *WebServer) getCachedPassword(keyBlob []byte, authToken string) ([]byte, error) {
+	_, pushes, err := encode.DecodeBlob(keyBlob)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding blob: %v", err)
+	}
```

Comment:
`if len(pushes) != 2 { ...` for good measure.

Reply:
Can't hurt.